store/sqlstore/channel_store.go

```diff
@@ -3100,3 +3352,272 @@ func (s SqlChannelStore) GroupSyncedChannelCount() (int64, *model.AppError) {
 
 	return count, nil
 }
+
+type sidebarCategoryForJoin struct {
+	model.SidebarCategory
+	ChannelId string
+}
+
+func (s SqlChannelStore) CreateSidebarCategory(userId, teamId string, newCategory *model.SidebarCategoryWithChannels) (*model.SidebarCategoryWithChannels, *model.AppError) {
+	transaction, err := s.GetMaster().Begin()
+	if err != nil {
+		return nil, model.NewAppError("SqlChannelStore.CreateSidebarCategory", "store.sql_channel.sidebar_categories.open_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	defer finalizeTransaction(transaction)
+
+	categoryId := model.NewId()
+	category := &model.SidebarCategory{
+		DisplayName: newCategory.DisplayName,
+		Id:          categoryId,
+		UserId:      userId,
+		TeamId:      teamId,
+		SortOrder:   0,
+		Type:        model.SidebarCategoryCustom,
+	}
+	if err = transaction.Insert(category); err != nil {
+		return nil, model.NewAppError("SqlPostStore.CreateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+	var channels []interface{}
+	runningOrder := 0
+	for _, channelID := range newCategory.Channels {
+		channels = append(channels, &model.SidebarChannel{
+			ChannelId:  channelID,
+			CategoryId: categoryId,
+			SortOrder:  int64(runningOrder),
+			UserId:     userId,
+		})
+		runningOrder += model.MinimalSidebarSortDistance
+	}
+	if err = transaction.Insert(channels...); err != nil {
+		return nil, model.NewAppError("SqlPostStore.CreateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	if err = transaction.Commit(); err != nil {
+		return nil, model.NewAppError("SqlChannelStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.commit_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	return &model.SidebarCategoryWithChannels{
+		SidebarCategory: *category,
+		Channels:        newCategory.Channels,
+	}, nil
+}
+
+func (s SqlChannelStore) GetSidebarCategory(userId, teamId, categoryId string) (*model.SidebarCategoryWithChannels, *model.AppError) {
+	var categories []*sidebarCategoryForJoin
+	sql, args, _ := s.getQueryBuilder().
+		Select("*", "SidebarChannels.ChannelId").
+		From("SidebarCategories, SidebarChannels").
+		Where(sq.And{
+			sq.Eq{"SidebarCategories.UserId": userId},
+			sq.Eq{"SidebarCategories.TeamId": teamId},
```

Comment:
I'm a bit confused by this - can a sidebar category be shared by multiple users, or by multiple teams of the same user? It looks like not from the model structs, but maybe I misunderstood the schema? I guess I just don't really understand what's going on with all these IDs on the category object.

Reply:
will fix